test/spinxsk/spinxsk.c

```diff
@@ -1992,6 +2022,182 @@ ProcessPkts(
     return TRUE;
 }
 
+#define MAX_HEADER_STORAGE TCP_HEADER_STORAGE
+
+typedef union _INET_ADDR_STORAGE {
+    IN_ADDR Ipv4;
+    IN6_ADDR Ipv6;
+} INET_ADDR_STORAGE;
+
+//
+// Build and inject frames in the receive path through the FNMP driver.
+//
+UINT8
+InjectFnmpRxPacket(_In_ FNMP_HANDLE Fnmp) {
+    const ETHERNET_ADDRESS localMac = FNMP_LOCAL_ETHERNET_ADDRESS_INIT;
+    const ETHERNET_ADDRESS remoteMac = FNMP_NEIGHBOR_ETHERNET_ADDRESS_INIT;
+
+    const UINT32 payloadBufferSize = 64'000;
+    UCHAR* payload = calloc(payloadBufferSize, sizeof(UCHAR));
+    FillRandomBuffer(payloadBufferSize, payload);
+
+    const ULONG frameBufferSize = FNMP_FRAME_MAX_BACKFILL + MAX_HEADER_STORAGE + payloadBufferSize;
+    UCHAR* frame = calloc(frameBufferSize, sizeof(UCHAR));
```

Comment:
please correct me if i'm wrong, but AFAIK, in .c files, XDP is fairly strict about declaring all variables prior to any non-declaration lines. we do call trivial functions while initializing variables, though.

```
INT x; // ok
INT y = GetCurrentTime(); // ok
x = y; // ok
INT z; // this is not ok.

Reply:
in c++, we use declare-as-you go, because it much harder to make use of c++ features with old-c-style declarations.

Reply:
XDP seems fairly consistent about it, spinxsk somewhat less (variables declared in loops, etc...)
On the other hand, what benefit do we get from declaring variables earlier than needed? It was required in very old C version - and sometimes is still needed because of goto-based error handling.
But generally, why put variables in a bigger scope than needed? Keeping the scope small helps with not accessing the variable in an unintended way (same logic as why global variables are generally discouraged)

Reply:
to be clear, the rules apply per scope, so the block within an `if` or `for` block can start declaring new variables.

as to why, it is simply the style.